cmd/jujud/updateseries/updateseries.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package updateseries
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/arch"
+	"github.com/juju/utils/fs"
+	jujuos "github.com/juju/utils/os"
+	"github.com/juju/utils/series"
+	"github.com/juju/utils/shell"
+	"github.com/juju/utils/symlink"
+	"github.com/juju/version"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/agent/tools"
+	"github.com/juju/juju/cmd/jujud/agent"
+	cmdutil "github.com/juju/juju/cmd/jujud/util"
+	"github.com/juju/juju/juju/paths"
+	"github.com/juju/juju/mongo"
+	"github.com/juju/juju/service"
+	"github.com/juju/juju/service/common"
+	"github.com/juju/juju/service/systemd"
+)
+
+var logger = loggo.GetLogger("juju.cmd.jujud.updateseries")
+
+const updateSeriesCommandDoc = `
+Update Juju agents running on this machine to start with Systemd.
+`
+
+type UpdateSeriesCommand struct {
+	cmd.CommandBase
+
+	machineAgent string
+	unitAgents   []string
+	dataDir      string
+	toSeries     string
+	fromSeries   string
+	startAgents  bool
+}
+
+func (c *UpdateSeriesCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "juju-updateseries",
+		Args:    "--to-series <series> --from-series <series> [--data-dir <dir>|--start-agents]",
+		Purpose: "Update Juju agents running on this machine to start with Systemd",
+		Doc:     updateSeriesCommandDoc,
+	}
+}
+
+func (c *UpdateSeriesCommand) Init(args []string) error {
+	switch {
+	case c.toSeries == "" && c.fromSeries == "":
+		return errors.Errorf("both --to-series and --from-series must be specified")
+	case c.toSeries == "":
+		return errors.Errorf("--to-series must be specified")
+	case c.fromSeries == "":
+		return errors.Errorf("--from-series must be specified")
+	case c.toSeries == c.fromSeries:
+		return errors.Errorf("--to-series and --from-series cannot be the same")
+	}
+
+	fromOS, err1 := series.GetOSFromSeries(c.toSeries)
+	toOS, err2 := series.GetOSFromSeries(c.fromSeries)
+	switch {
+	case err1 != nil:
+		return err1
+	case err2 != nil:
+		return err2
+	case fromOS != toOS:
+		return errors.Errorf("series from two different operating systems specified")
+	case fromOS == jujuos.Windows:
+		return errors.NewNotSupported(nil, "windows not supported")
+	}
+
+	ctlr, err := isController()
+	switch {
+	case err != nil:
+		return err
+	case ctlr:
+		return errors.Errorf("cannot run on a controller machine")
```

Comment:
I guess we don't want the underlying controller os to change. Openstack bundle does deploy in models on the controller machine as part of default config I believe. Would a setup like that still work?

Reply:
If you update the series of a controller, mongo fails horribly.  We don't want deal with that.  The strategy for updating a controller is to bootstrap a new one and migrate your models.

The openstack bundles I've seen don't deploy the a controller machine??